lisa/sut_orchestrator/qemu/platform.py

```diff
@@ -103,7 +103,10 @@ def _deploy_nodes(self, environment: Environment, log: Logger) -> None:
                 self._fill_nodes_metadata(environment, log, qemu_conn)
 
             except Exception as ex:
-                self._delete_nodes(environment, log, qemu_conn)
+                assert environment.platform
+                if not environment.platform.runbook.keep_environment:
```

Comment:
It's not a bool value, when code reaches here. It has three values: always, no, failed. The constants are defined in constatns.py. ENVIRONMENT_KEEP_ALWAYS/NO/FAILED.